evm-ds/src/main.rs

```diff
@@ -98,44 +93,123 @@ impl CallContext {
+            pc: Default::default(),
+            stack: Default::default(),
+            storage: Default::default(),
+        }
+    }
+}
 
 // This implementation has a stack of call contexts each with reference to their calls - so a tree is
 // Created in this way.
 // Each new call gets added to the end of the stack and becomes the current context.
 // On returning from a call, the end of the stack gets put into the item above's calls
 #[derive(Debug,Serialize,Deserialize)]
 struct LoggingEventListener {
-    call_stack: Vec<CallContext>,
+    call_tracer: Vec<CallContext>,
+    raw_tracer: StructLogTopLevel,
     enabled: bool,
 }
 
-impl LoggingEventListener {
+#[derive(Debug,Serialize,Deserialize, Default)]
+struct StructLogTopLevel {
+    pub gas: u64,
+    #[serde(rename = "returnValue")]
+    pub return_value: String,
+    #[serde(rename = "structLogs")]
+    pub struct_logs: Vec<StructLog>,
+}
 
+impl LoggingEventListener {
     fn new(enabled: bool) -> Self {
         LoggingEventListener {
-            call_stack: Default::default(),
+            call_tracer: Default::default(),
+            raw_tracer: Default::default(),
             enabled: enabled,
         }
     }
 }
 
-impl tracing::EventListener for LoggingEventListener {
-    fn event(&mut self, event: tracing::Event) {
-        println!("recvd event: {:?}", event);
+impl evm::runtime::tracing::EventListener for LoggingEventListener {
+    fn event(&mut self, event: evm::runtime::tracing::Event) {
+
+        if !self.enabled {
+            return;
+        }
+
+        let mut struct_log = StructLog::new();
+
+        struct_log.depth = self.call_tracer.len() - 1;
+        struct_log.error = Default::default();
+        struct_log.gas = Default::default();
+        struct_log.gas_cost = Default::default();
+        struct_log.storage = Default::default();
+
+        match event {
+            evm::runtime::tracing::Event::Step{context: _, opcode, position, stack, memory: _} => {
+                struct_log.op = format!("{}", opcode);
+                struct_log.pc = position.clone().unwrap_or(0);
+
+                for sta in stack.data() {
+                    struct_log.stack.push(format!("{:?}", sta));
+                }
+            }
+            evm::runtime::tracing::Event::StepResult{result, return_value: _} => {
+                struct_log.op = "StepResult".to_string();
+                struct_log.error = format!("{:?}", result.clone());
+            }
+            evm::runtime::tracing::Event::SLoad{address: _, index: _, value: _} => {
+                struct_log.op = "Sload".to_string();
+            }
+            evm::runtime::tracing::Event::SStore{address: _, index: _, value: _} => {
+                struct_log.op = "SStore".to_string();
+            }
+        }
+
+        if self.raw_tracer.struct_logs.len() < 5 {
+            self.raw_tracer.struct_logs.push(struct_log);
+        }
     }
 }
 
 impl LoggingEventListener {
 
+    #[allow(dead_code)]
```

Comment:
If it is dead code, do we need it?

Reply:
I like when debugging to switch to pretty strings which is otherwise unused